File: 백준/Silver/1966. 프린터 큐/프린터 큐.cc
#include <iostream>
#include <queue>
using namespace std;
int main() {
	int cnt;
	int T;
	cin >> T;
	int n, m;
	int k;
	while (T--) {
		cnt = 0;
		cin >> n >> m;
		queue<pair<int, int>>q;
		priority_queue<int>pq;
		for (int j = 0; j < n; j++) {
			cin >> k;
			q.push({ j,k });
			pq.push(k);
		}
		while (!q.empty()) {
			int idx = q.front().first;
			int val = q.front().second;
			q.pop();
			if (pq.top() == val) {
				pq.pop();
				cnt++;
				if (idx == m) {
					cout << cnt << endl;
					break;
				}
			}
			else q.push({ idx,val });
		}
	}
}

File: 백준/Silver/1966. 프린터 큐/프린터 큐.java
import java.util.*;

public class Main {

    public static class Node{
        int val;
        int idx;
        public Node(int val, int idx){
            this.val = val;
            this.idx = idx;
        }
    }

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        int T = sc.nextInt();

        for(int i=0;i<T;i++){
            int n = sc.nextInt();
            int m = sc.nextInt();
            Queue<Node> queue = new ArrayDeque<>();
            Queue<Node> pq = new PriorityQueue<>((a,b)->Integer.compare(b.val,a.val));

            for(int j=0;j<n;j++){
                int x = sc.nextInt();
                queue.add(new Node(x,j));
                pq.add(new Node(x,j));
            }

            int cnt = 1;
            while(!queue.isEmpty()){
                Node curr = queue.poll();
                if(curr.idx == m && pq.peek().val == curr.val){
                    System.out.println(cnt);
                    break;
                }else if(pq.peek().val == curr.val){
                    pq.poll();
                    cnt++;
                }else{
                    queue.add(curr);
                }
            }
        }
    }
}